### PR TITLE
Switch the keadm ctl client to client go and add the restful interface for restart pod.

### PR DESCRIPTION
**Pull request by luomengY**
…e for restart pod.

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
1.switch the keadm ctl client to client go.
2.add the restful interface for restart pod.
POST:/api/v1/namespaces/{namespace}/pods/restart
body: ["pod-1","pod-2","pod-3"]
response: {"errMessages":[],"containerIds":[]}
image

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: keadm/cmd/keadm/app/cmd/ctl/restart/pod.go [hunk 1]
```diff
@@ -71,51 +69,57 @@ func AddRestartPodFlags(cmd *cobra.Command, RestartPodOptions *PodRestartOptions
...
+		return err
+	}
+
+	for _, errMsg := range restartResponse.ErrMessages {
+		fmt.Println(errMsg)
+	}
+
+	for _, containerID := range restartResponse.ContainerIds {
+		fmt.Println(containerID)
 	}
```
Comment (WillardHu) [inline keadm/cmd/keadm/app/cmd/ctl/restart/pod.go:87]: That's not a friendly way to output messages, right?

File: keadm/cmd/keadm/app/cmd/ctl/restart/pod.go [hunk 2]
```diff
@@ -71,51 +69,57 @@ func AddRestartPodFlags(cmd *cobra.Command, RestartPodOptions *PodRestartOptions
...
 	return nil
 }
+
+func podRestart(clientSet *kubernetes.Clientset, namespace string, podNames []string) (*storage.RestartResponse, error) {
+	bodyBytes, err := json.Marshal(podNames)
+	if err != nil {
+		return nil, err
+	}
+	ctx := context.TODO()
```
Comment (WillardHu) [inline keadm/cmd/keadm/app/cmd/ctl/restart/pod.go:96]: Is it better if the variable context is inputed by the caller?

File: keadm/cmd/keadm/app/cmd/ctl/client/pod.go [hunk 3]
```diff
@@ -30,42 +31,36 @@ type PodRequest struct {
...
-		request = Request{
-			Method: "GET",
-			Path:   "/" + CoreAPIPrefix + "/" + CoreAPIGroupVersion.Version + "/pods",
-		}
-	} else {
-		request = Request{
-			Method: "GET",
-			Path: "/" + CoreAPIPrefix + "/" + CoreAPIGroupVersion.Version +
-				"/namespaces/" + podRequest.Namespace + "/pods",
+		podList, err := kubeClient.CoreV1().Pods(metaV1.NamespaceAll).List(context.TODO(), metaV1.ListOptions{
```
Comment (WillardHu) [inline keadm/cmd/keadm/app/cmd/ctl/client/pod.go:52]: Is it better if the variable context is inputed by the caller?

Comment (luomengY) [reply keadm/cmd/keadm/app/cmd/ctl/client/pod.go]: ok

File: keadm/cmd/keadm/app/cmd/ctl/restart/pod.go (hunk 2 above)
Comment (luomengY) [reply keadm/cmd/keadm/app/cmd/ctl/restart/pod.go]: ok

File: keadm/cmd/keadm/app/cmd/ctl/restart/pod.go (hunk 1 above)
Comment (luomengY) [reply keadm/cmd/keadm/app/cmd/ctl/restart/pod.go]: > That's not a friendly way to output messages, right?

What is the appropriate way to print the stopped container ID and the error message returned by the request here?

Comment (WillardHu) [reply keadm/cmd/keadm/app/cmd/ctl/restart/pod.go]: Follow your logic, friendly information sample:
```
case 1. The pod 'aaa' restart failure, reason: <error-message>.
case 2. There are some containers of Pod 'bbb' that failed to restart, reason:  <error-message>,  containers container-1, container-2 have been restarted.
case 3. The pod 'ccc' restart successful.
```
You can handle these response error messages in package storage.
```golang
if err != nil {
    message := fmt.Sprintf(xxxxxxxxx)
    klog.Warning("[metaserver/reststorage] ", message)
    restartResponse.ErrMessages = append(restartResponse.ErrMessages, message)
}

```

Comment (luomengY) [reply keadm/cmd/keadm/app/cmd/ctl/restart/pod.go]: > Follow your logic, friendly information sample:
> 
> ```
> case 1. The pod 'aaa' restart failure, reason: <error-message>.
> case 2. There are some containers of Pod 'bbb' that failed to restart, reason:  <error-message>,  containers container-1, container-2 have been restarted.
> case 3. The pod 'ccc' restart successful.
> ```
> 
> You can handle these response error messages in package storage.
> 
> ```go
> if err != nil {
>     message := fmt.Sprintf(xxxxxxxxx)
>     klog.Warning("[metaserver/reststorage] ", message)
>     restartResponse.ErrMessages = append(restartResponse.ErrMessages, message)
> }
> ```

Done

File: keadm/cmd/keadm/app/cmd/ctl/client/pod.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+type PodRequest struct {
+	Namespace     string
+	LabelSelector string
+	AllNamespaces bool
+	PodName       string
+}
+
+func (podRequest *PodRequest) GetPod(ctx context.Context) (*corev1.Pod, error) {
+	select {
```
Comment (WillardHu) [inline keadm/cmd/keadm/app/cmd/ctl/client/pod.go:34]: Why use select here?

File: keadm/cmd/keadm/app/cmd/ctl/client/pod.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+		pod, err := kubeClient.CoreV1().Pods(podRequest.Namespace).Get(ctx, podRequest.PodName, metaV1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		return pod, nil
+	}
+}
+
+func (podRequest *PodRequest) GetPods(ctx context.Context) (*corev1.PodList, error) {
+	select {
```
Comment (WillardHu) [inline keadm/cmd/keadm/app/cmd/ctl/client/pod.go:51]: select?

File: keadm/cmd/keadm/app/cmd/ctl/client/pod.go (hunk 4 above)
Comment (luomengY) [reply keadm/cmd/keadm/app/cmd/ctl/client/pod.go]: > Why use select here?

Modified

File: keadm/cmd/keadm/app/cmd/ctl/client/pod.go (hunk 5 above)
Comment (luomengY) [reply keadm/cmd/keadm/app/cmd/ctl/client/pod.go]: > select?

Modified

File: edge/pkg/metamanager/metaserver/handlerfactory/handler.go [hunk 6]
```diff
@@ -200,6 +201,46 @@ func (f *Factory) Patch(reqInfo *request.RequestInfo) http.Handler {
 	return http.HandlerFunc(h)
 }
 
+func (f *Factory) Restart(namespace string) http.Handler {
```
[the remark this answers is not in the document]
Comment (luomengY) [reply edge/pkg/metamanager/metaserver/handlerfactory/handler.go]: > ext_pods_handler.go

ok

File: edge/pkg/metamanager/metaserver/server.go [hunk 7]
```diff
@@ -183,7 +183,11 @@ func (ls *MetaServer) BuildBasicHandler() http.Handler {
 			case reqInfo.Verb == "list", reqInfo.Verb == "watch":
 				ls.Factory.List().ServeHTTP(w, req)
 			case reqInfo.Verb == "create":
-				ls.Factory.Create(reqInfo).ServeHTTP(w, req)
+				if reqInfo.Name == "restart" {
```
[the remark this answers is not in the document]
Comment (luomengY) [reply edge/pkg/metamanager/metaserver/server.go:186]: > `if reqInfo.SubResource == "restart"`?

The Subresource here is nil, Name is "restart".

File: pkg/metaserver/constants.go [hunk 8]
```diff
@@ -59,6 +59,11 @@ type PatchInfo struct {
 	Subresources []string
 }
 
+type RestartInfo struct {
```
Comment (WillardHu) [inline pkg/metaserver/constants.go:62]: Defining structures in the constants.go is not good, you can add a file types.go to maintain them, the struct `RestartResponse` can also be moved over to this file, it is not good that the keadm module depend on the metaserver module package.

Comment (luomengY) [reply pkg/metaserver/constants.go]: Done

**Comment by luomengY** [comment]
@fisherxu

**Comment by luomengY** [comment]
PTAL @Shelley-BaoYue

**Comment by luomengY** [comment]
PTAL @Shelley-BaoYue @WillardHu

**Pull request merged**